internal/lsp/completions/providers/packagename.go

```diff
@@ -0,0 +1,70 @@
+package providers
+
+import (
+	"fmt"
+	"path/filepath"
+	"strings"
+
+	"github.com/styrainc/regal/internal/lsp/cache"
+	"github.com/styrainc/regal/internal/lsp/clients"
+	"github.com/styrainc/regal/internal/lsp/types"
+	"github.com/styrainc/regal/internal/lsp/uri"
+)
+
+// PackageName will return completions for the package name when starting a new file based on the file's URI.
+type PackageName struct{}
+
+func (*PackageName) Run(c *cache.Cache, params types.CompletionParams) ([]types.CompletionItem, error) {
+	fileURI := params.TextDocument.URI
+
+	fileContents, ok := c.GetFileContents(fileURI)
+	if !ok {
+		// if the file contents is missing then we can't provide completions
+		return nil, nil
+	}
+
+	lines := strings.Split(fileContents, "\n")
+	if len(lines) < 1 {
+		return nil, nil
+	}
+
+	for i, line := range lines {
+		if i < int(params.Position.Line) && strings.HasPrefix(line, "package ") {
+			// if there is already a package statement in the file then we don't provide any more completions
+			return nil, nil
+		}
+	}
+
+	// if we can't confirm that the user has package statement on the line then we don't provide completions
+	if len(lines)+1 < int(params.Position.Line) {
+		return nil, nil
+	}
+
+	path := uri.ToPath(clients.IdentifierGeneric, fileURI)
+	dir := filepath.Base(filepath.Dir(path))
```

Comment:
One of the challenges here is that we'd need to know where to stop, this would require being told the workspace root. should we always suggest the longest path, or all three? e/g/ bugs, rules.bugs, and regal.rules.bugs?

Reply:
Yeah, that is indeed a challenge. I guess what we could do is to start from the first 3 (or so) characters that match, so if you have something like:

```
/Users/charlie/projects/regal/rules/bugs
```

And you typed in `package reg` it would base the suggestion from the first matching position in the string, which would be `regal/rules/bugs` in this case. But yeah, I'm sure there would be cases where it could get tricky.

Reply:
Using the workspace root from the server's config feels like the right option here.

Do you think that we'd always want to suggest the longest package name we can, and that one only? I.e. only suggest `regal.rules.bugs`


If so, that might be worth doing as part of this PR.

Reply:
<img width="880" alt="Screenshot 2024-05-16 at 11 18 39" src="https://github.com/StyraInc/regal/assets/1774239/5327f3d8-768d-4f68-9fd9-8e1a763331a4">


See c163642ec742701a0866d417140003278b5f957e

Reply:
very nice!